<?xml version="1.0" encoding="UTF-8"?>
<ui version="4.0">
 <class>optSetupFrame</class>
 <widget class="QFrame" name="optSetupFrame">
  <property name="geometry">
   <rect>
    <x>0</x>
    <y>0</y>
    <width>1063</width>
    <height>777</height>
   </rect>
  </property>
  <property name="cursor">
   <cursorShape>ArrowCursor</cursorShape>
  </property>
  <property name="windowTitle">
   <string>Frame</string>
  </property>
  <property name="frameShape">
   <enum>QFrame::StyledPanel</enum>
  </property>
  <property name="frameShadow">
   <enum>QFrame::Raised</enum>
  </property>
  <layout class="QVBoxLayout" name="verticalLayout_9">
   <item>
    <widget class="QTabWidget" name="tabWidget">
     <property name="cursor">
      <cursorShape>PointingHandCursor</cursorShape>
     </property>
     <property name="currentIndex">
      <number>0</number>
     </property>
     <widget class="QWidget" name="tab">
      <attribute name="title">
       <string>Variables</string>
      </attribute>
      <layout class="QVBoxLayout" name="verticalLayout_3">
       <item>
        <widget class="QTableWidget" name="varForm">
         <property name="sizePolicy">
          <sizepolicy hsizetype="Expanding" vsizetype="Expanding">
           <horstretch>0</horstretch>
           <verstretch>0</verstretch>
          </sizepolicy>
         </property>
         <property name="minimumSize">
          <size>
           <width>5</width>
           <height>5</height>
          </size>
         </property>
         <column>
          <property name="text">
           <string>Variable</string>
          </property>
         </column>
         <column>
          <property name="text">
           <string>Type</string>
          </property>
         </column>
         <column>
          <property name="text">
           <string>Scale</string>
          </property>
         </column>
         <column>
          <property name="text">
           <string>Min</string>
          </property>
         </column>
         <column>
          <property name="text">
           <string>Max</string>
          </property>
         </column>
         <column>
          <property name="text">
           <string>Value</string>
          </property>
         </column>
        </widget>
       </item>
      </layout>
     </widget>
     <widget class="QWidget" name="tab_4">
      <attribute name="title">
       <string>Samples</string>
      </attribute>
      <layout class="QVBoxLayout" name="verticalLayout">
       <item>
        <layout class="QHBoxLayout" name="horizontalLayout_5">
         <item>
          <widget class="QPushButton" name="addSampleButton">
           <property name="text">
            <string>Add Sample</string>
           </property>
           <property name="icon">
            <iconset resource="../icons.qrc">
             <normaloff>:/icons/icons/add.svg</normaloff>:/icons/icons/add.svg</iconset>
           </property>
          </widget>
         </item>
         <item>
          <widget class="QPushButton" name="deleteSampleButton">
           <property name="text">
            <string>Delete Samples</string>
           </property>
           <property name="icon">
            <iconset resource="../icons.qrc">
             <normaloff>:/icons/icons/delete.svg</normaloff>:/icons/icons/delete.svg</iconset>
           </property>
          </widget>
         </item>
         <item>
          <widget class="QPushButton" name="generateSamplesButton">
           <property name="text">
            <string>Generate Samples</string>
           </property>
           <property name="icon">
            <iconset resource="../icons.qrc">
             <normaloff>:/icons/icons/down.svg</normaloff>:/icons/icons/down.svg</iconset>
           </property>
          </widget>
         </item>
         <item>
          <widget class="QPushButton" name="clearSamplesButton">
           <property name="text">
            <string>Clear Samples</string>
           </property>
           <property name="icon">
            <iconset resource="../icons.qrc">
             <normaloff>:/icons/icons/new.svg</normaloff>:/icons/icons/new.svg</iconset>
           </property>
          </widget>
         </item>
         <item>
          <spacer name="horizontalSpacer_4">
           <property name="orientation">
            <enum>Qt::Horizontal</enum>
           </property>
           <property name="sizeHint" stdset="0">
            <size>
             <width>40</width>
             <height>20</height>
            </size>
           </property>
          </spacer>
         </item>
        </layout>
       </item>
       <item>
        <widget class="QTableWidget" name="sampleTable">
         <property name="palette">
          <palette>
           <active>
            <colorrole role="AlternateBase">
             <brush brushstyle="SolidPattern">
              <color alpha="255">
               <red>222</red>
               <green>255</green>
               <blue>222</blue>
              </color>
             </brush>
            </colorrole>
           </active>
           <inactive>
            <colorrole role="AlternateBase">
             <brush brushstyle="SolidPattern">
              <color alpha="255">
               <red>222</red>
               <green>255</green>
               <blue>222</blue>
              </color>
             </brush>
            </colorrole>
           </inactive>
           <disabled>
            <colorrole role="AlternateBase">
             <brush brushstyle="SolidPattern">
              <color alpha="255">
               <red>222</red>
               <green>255</green>
               <blue>222</blue>
              </color>
             </brush>
            </colorrole>
           </disabled>
          </palette>
         </property>
         <property name="alternatingRowColors">
          <bool>true</bool>
         </property>
        </widget>
       </item>
      </layout>
     </widget>
     <widget class="QWidget" name="tab_3">
      <attribute name="title">
       <string>Objective/Constraints</string>
      </attribute>
      <layout class="QVBoxLayout" name="verticalLayout_10">
       <item>
        <layout class="QHBoxLayout" name="horizontalLayout_6">
         <item>
          <widget class="QComboBox" name="objTypeCombo">
           <property name="minimumSize">
            <size>
             <width>300</width>
             <height>0</height>
            </size>
           </property>
           <item>
            <property name="text">
             <string>Simple Python Expression</string>
            </property>
           </item>
           <item>
            <property name="text">
             <string>Custom Python</string>
            </property>
           </item>
          </widget>
         </item>
         <item>
          <spacer name="horizontalSpacer_5">
           <property name="orientation">
            <enum>Qt::Horizontal</enum>
           </property>
           <property name="sizeHint" stdset="0">
            <size>
             <width>40</width>
             <height>20</height>
            </size>
           </property>
          </spacer>
         </item>
        </layout>
       </item>
       <item>
        <widget class="QStackedWidget" name="stackedWidget">
         <property name="currentIndex">
          <number>0</number>
         </property>
         <widget class="QWidget" name="page">
          <layout class="QVBoxLayout" name="verticalLayout_11">
           <item>
            <widget class="QGroupBox" name="groupBox">
             <property name="sizePolicy">
              <sizepolicy hsizetype="Preferred" vsizetype="Preferred">
               <horstretch>0</horstretch>
               <verstretch>1</verstretch>
              </sizepolicy>
             </property>
             <property name="minimumSize">
              <size>
               <width>5</width>
               <height>5</height>
              </size>
             </property>
             <property name="title">
              <string>Objective Function f(x)</string>
             </property>
             <layout class="QHBoxLayout" name="horizontalLayout_2">
              <item>
               <widget class="QTableWidget" name="fTable">
                <property name="minimumSize">
                 <size>
                  <width>5</width>
                  <height>5</height>
                 </size>
                </property>
                <column>
                 <property name="text">
                  <string>Expression</string>
                 </property>
                </column>
                <column>
                 <property name="text">
                  <string>Penalty Scale</string>
                 </property>
                </column>
                <column>
                 <property name="text">
                  <string>Value for Failure</string>
                 </property>
                </column>
               </widget>
              </item>
              <item>
               <layout class="QVBoxLayout" name="verticalLayout_5">
                <item>
                 <widget class="QPushButton" name="fAddButton">
                  <property name="minimumSize">
                   <size>
                    <width>0</width>
                    <height>0</height>
                   </size>
                  </property>
                  <property name="cursor">
                   <cursorShape>PointingHandCursor</cursorShape>
                  </property>
                  <property name="text">
                   <string/>
                  </property>
                  <property name="icon">
                   <iconset>
                    <normalon>:/icons/icons/add.svg</normalon>
                   </iconset>
                  </property>
                  <property name="iconSize">
                   <size>
                    <width>16</width>
                    <height>16</height>
                   </size>
                  </property>
                  <property name="flat">
                   <bool>false</bool>
                  </property>
                 </widget>
                </item>
                <item>
                 <widget class="QPushButton" name="fDelButton">
                  <property name="minimumSize">
                   <size>
                    <width>5</width>
                    <height>5</height>
                   </size>
                  </property>
                  <property name="cursor">
                   <cursorShape>PointingHandCursor</cursorShape>
                  </property>
                  <property name="text">
                   <string/>
                  </property>
                  <property name="icon">
                   <iconset>
                    <normalon>:/icons/icons/remove.svg</normalon>
                   </iconset>
                  </property>
                  <property name="iconSize">
                   <size>
                    <width>16</width>
                    <height>16</height>
                   </size>
                  </property>
                  <property name="flat">
                   <bool>false</bool>
                  </property>
                 </widget>
                </item>
                <item>
                 <spacer name="verticalSpacer">
                  <property name="orientation">
                   <enum>Qt::Vertical</enum>
                  </property>
                  <property name="sizeHint" stdset="0">
                   <size>
                    <width>20</width>
                    <height>40</height>
                   </size>
                  </property>
                 </spacer>
                </item>
               </layout>
              </item>
             </layout>
            </widget>
           </item>
           <item>
            <widget class="QGroupBox" name="groupBox_2">
             <property name="sizePolicy">
              <sizepolicy hsizetype="Preferred" vsizetype="Preferred">
               <horstretch>0</horstretch>
               <verstretch>1</verstretch>
              </sizepolicy>
             </property>
             <property name="minimumSize">
              <size>
               <width>5</width>
               <height>5</height>
              </size>
             </property>
             <property name="title">
              <string>Inequality Constraints g(x) &lt;= 0</string>
             </property>
             <layout class="QHBoxLayout" name="horizontalLayout_4">
              <item>
               <widget class="QTableWidget" name="gTable">
                <property name="minimumSize">
                 <size>
                  <width>5</width>
                  <height>5</height>
                 </size>
                </property>
                <column>
                 <property name="text">
                  <string>Expression</string>
                 </property>
                </column>
                <column>
                 <property name="text">
                  <string>Penalty Factor</string>
                 </property>
                </column>
                <column>
                 <property name="text">
                  <string>Form</string>
                 </property>
                </column>
               </widget>
              </item>
              <item>
               <layout class="QVBoxLayout" name="verticalLayout_2">
                <item>
                 <widget class="QPushButton" name="gAddButton">
                  <property name="minimumSize">
                   <size>
                    <width>0</width>
                    <height>0</height>
                   </size>
                  </property>
                  <property name="cursor">
                   <cursorShape>PointingHandCursor</cursorShape>
                  </property>
                  <property name="text">
                   <string/>
                  </property>
                  <property name="icon">
                   <iconset>
                    <normalon>:/icons/icons/add.svg</normalon>
                   </iconset>
                  </property>
                  <property name="iconSize">
                   <size>
                    <width>16</width>
                    <height>16</height>
                   </size>
                  </property>
                  <property name="flat">
                   <bool>false</bool>
                  </property>
                 </widget>
                </item>
                <item>
                 <widget class="QPushButton" name="gDelButton">
                  <property name="cursor">
                   <cursorShape>PointingHandCursor</cursorShape>
                  </property>
                  <property name="text">
                   <string/>
                  </property>
                  <property name="icon">
                   <iconset>
                    <normalon>:/icons/icons/remove.svg</normalon>
                   </iconset>
                  </property>
                  <property name="iconSize">
                   <size>
                    <width>16</width>
                    <height>16</height>
                   </size>
                  </property>
                  <property name="flat">
                   <bool>false</bool>
                  </property>
                 </widget>
                </item>
                <item>
                 <spacer name="verticalSpacer_2">
                  <property name="orientation">
                   <enum>Qt::Vertical</enum>
                  </property>
                  <property name="sizeHint" stdset="0">
                   <size>
                    <width>20</width>
                    <height>40</height>
                   </size>
                  </property>
                 </spacer>
                </item>
               </layout>
              </item>
             </layout>
            </widget>
           </item>
           <item>
            <layout class="QHBoxLayout" name="horizontalLayout">
             <item>
              <spacer name="horizontalSpacer_2">
               <property name="orientation">
                <enum>Qt::Horizontal</enum>
               </property>
               <property name="sizeHint" stdset="0">
                <size>
                 <width>40</width>
                 <height>20</height>
                </size>
               </property>
              </spacer>
             </item>
             <item>
              <widget class="QPushButton" name="checkInputButton">
               <property name="font">
                <font>
                 <pointsize>8</pointsize>
                 <underline>false</underline>
                </font>
               </property>
               <property name="cursor">
                <cursorShape>PointingHandCursor</cursorShape>
               </property>
               <property name="text">
                <string>Check Input...</string>
               </property>
               <property name="icon">
                <iconset>
                 <normalon>:/icons/icons/check.svg</normalon>
                </iconset>
               </property>
               <property name="iconSize">
                <size>
                 <width>16</width>
                 <height>16</height>
                </size>
               </property>
               <property name="flat">
                <bool>false</bool>
               </property>
              </widget>
             </item>
             <item>
              <widget class="QPushButton" name="vexButton">
               <property name="font">
                <font>
                 <pointsize>8</pointsize>
                 <underline>false</underline>
                </font>
               </property>
               <property name="cursor">
                <cursorShape>PointingHandCursor</cursorShape>
               </property>
               <property name="text">
                <string>Variable Explorer</string>
               </property>
               <property name="icon">
                <iconset>
                 <normalon>:/icons/icons/search.svg</normalon>
                </iconset>
               </property>
               <property name="iconSize">
                <size>
                 <width>16</width>
                 <height>16</height>
                </size>
               </property>
               <property name="flat">
                <bool>false</bool>
               </property>
              </widget>
             </item>
             <item>
              <spacer name="horizontalSpacer_3">
               <property name="orientation">
                <enum>Qt::Horizontal</enum>
               </property>
               <property name="sizeHint" stdset="0">
                <size>
                 <width>40</width>
                 <height>20</height>
                </size>
               </property>
              </spacer>
             </item>
            </layout>
           </item>
          </layout>
         </widget>
         <widget class="QWidget" name="page_4">
          <layout class="QVBoxLayout" name="verticalLayout_13">
           <item>
            <widget class="QGroupBox" name="groupBox_4">
             <property name="title">
              <string>Custom Python</string>
             </property>
             <layout class="QVBoxLayout" name="verticalLayout_12">
              <item>
               <widget class="QPlainTextEdit" name="plainTextEdit_2">
                <property name="maximumSize">
                 <size>
                  <width>16777215</width>
                  <height>150</height>
                 </size>
                </property>
                <property name="font">
                 <font>
                  <family>Courier</family>
                 </font>
                </property>
                <property name="readOnly">
                 <bool>true</bool>
                </property>
                <property name="plainText">
                 <string>Enter Python below to define an objective function.  The function should return three things the first is the objective function list.  This is a list to support multi-objective problems.  The other two items are only diagnostic, they are not use in any calulation.  The second item is a list of constraint violations.  The lst item is the constraint penalty.  In custom mode it is up to the user to apply constraint penalties.

The arguments to the function are x, f, and fail.  The inputs and outputs of a flowsheet run are stored in x and f.  If sampling is used the first index is the sample index, and the second key is the node name.  If there is one sample only, the first key is the node a name.  The next key is the variables name, and the last index is the time step index.  The timestep index would usually be 0.  Fail is a boolean list.  It indicates if any flowsheet sample failed.

Example:

def objfunc(x, f, fail):
    # Simple example where the objective is just a flowsheet output
    if any(fail): #simulation failed to converge
        obj = [100000]
    else: #simulation successful
        obj = [f['Node']['x'][0]]
    return obj, [0], 0</string>
                </property>
               </widget>
              </item>
              <item>
               <widget class="QPlainTextEdit" name="customCodeEdit">
                <property name="font">
                 <font>
                  <family>Courier</family>
                  <pointsize>10</pointsize>
                 </font>
                </property>
               </widget>
              </item>
             </layout>
            </widget>
           </item>
          </layout>
         </widget>
        </widget>
       </item>
      </layout>
     </widget>
     <widget class="QWidget" name="tab_2">
      <attribute name="title">
       <string>Solver</string>
      </attribute>
      <layout class="QVBoxLayout" name="verticalLayout_7">
       <item>
        <widget class="QGroupBox" name="groupBox_5">
         <property name="sizePolicy">
          <sizepolicy hsizetype="Preferred" vsizetype="Maximum">
           <horstretch>0</horstretch>
           <verstretch>0</verstretch>
          </sizepolicy>
         </property>
         <property name="title">
          <string>Solver</string>
         </property>
         <layout class="QVBoxLayout" name="verticalLayout_6">
          <item>
           <layout class="QHBoxLayout" name="horizontalLayout_3">
            <item>
             <widget class="QLabel" name="label_3">
              <property name="text">
               <string>Select Solver</string>
              </property>
              <property name="buddy">
               <cstring>solverBox</cstring>
              </property>
             </widget>
            </item>
            <item>
             <widget class="QComboBox" name="solverBox">
              <property name="sizePolicy">
               <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
                <horstretch>0</horstretch>
                <verstretch>0</verstretch>
               </sizepolicy>
              </property>
             </widget>
            </item>
            <item>
             <spacer name="horizontalSpacer">
              <property name="orientation">
               <enum>Qt::Horizontal</enum>
              </property>
              <property name="sizeHint" stdset="0">
               <size>
                <width>40</width>
                <height>20</height>
               </size>
              </property>
             </spacer>
            </item>
           </layout>
          </item>
          <item>
           <widget class="QGroupBox" name="groupBox_3">
            <property name="maximumSize">
             <size>
              <width>16777215</width>
              <height>150</height>
             </size>
            </property>
            <property name="title">
             <string>Description</string>
            </property>
            <layout class="QVBoxLayout" name="verticalLayout_4">
             <item>
              <widget class="QTextBrowser" name="methodDescriptionBox"/>
             </item>
            </layout>
           </widget>
          </item>
         </layout>
        </widget>
       </item>
       <item>
        <widget class="QGroupBox" name="groupBox_6">
         <property name="minimumSize">
          <size>
           <width>5</width>
           <height>5</height>
          </size>
         </property>
         <property name="title">
          <string>Solver Options</string>
         </property>
         <layout class="QVBoxLayout" name="verticalLayout_8">
          <item>
           <widget class="QTableWidget" name="solverOptionTable">
            <property name="minimumSize">
             <size>
              <width>5</width>
              <height>5</height>
             </size>
            </property>
            <column>
             <property name="text">
              <string>Option</string>
             </property>
            </column>
            <column>
             <property name="text">
              <string>Setting</string>
             </property>
            </column>
            <column>
             <property name="text">
              <string>Description</string>
             </property>
            </column>
           </widget>
          </item>
         </layout>
        </widget>
       </item>
      </layout>
     </widget>
    </widget>
   </item>
  </layout>
 </widget>
 <resources>
  <include location="foqus_lib.gui.icons.qrc"/>
 </resources>
 <connections>
  <connection>
   <sender>objTypeCombo</sender>
   <signal>currentIndexChanged(int)</signal>
   <receiver>stackedWidget</receiver>
   <slot>setCurrentIndex(int)</slot>
   <hints>
    <hint type="sourcelabel">
     <x>237</x>
     <y>53</y>
    </hint>
    <hint type="destinationlabel">
     <x>1020</x>
     <y>76</y>
    </hint>
   </hints>
  </connection>
 </connections>
</ui>
